Refactor for code readability

MarkZH opened the pull request:
These changes focus on code readability through the following changes:

- Use named variables instead of in-line expressions (see `draw_or_resign()` in lichess-bot.py)
- Changing logic flows to reduced repeated code (see `get_lichess_cloud_move()` in lichess-bot.py)
- More efficient use of `dict.get()` with default arguments (see `Challenge.__init__()` in model.py)
- Creating functions to separate in-line code (see challenge declining in lichess-bot.py)
- Use triple-quoted strings for long multi-line strings (see the Stockfish strategy in `test-bot.py`)
- As a last resort, splitting function arguments across multiple lines

The flake8 warnings are much reduced

MarkZH commented:
I realize this is a huge pull request. Let me know if you want it split up--perhaps by file.

AttackingOrDefending commented on test_bot/lichess.py:
```diff
@@ -54,11 +86,20 @@ def iter_lines(self):
-                yield eval(f'b\'{{"type":"gameState","moves":"{moves}","wtime":{int(wtime * 1000)},"btime":{int(btime * 1000)},"winc":2000,"binc":2000,"status":"outoftime","winner":"black"}}\'')
+                new_game_state["status"] = "outoftime"
+                new_game_state["winner"] = "black"
+                yield bytes(str(new_game_state).replace("'", '"'), 'utf-8')
```
```python
json.dumps(new_game_state)
```
can probably be used instead of
```python
str(new_game_state).replace("'", '"')
```
I think the code looks cleaner that way, even though we are importing one more module.

AttackingOrDefending commented on test_bot/lichess.py:
```diff
@@ -150,7 +197,15 @@ def decline_challenge(self, challenge_id, reason="generic"):
+                   "username": "b",
+                   "online": True,
+                   "title": "BOT",
+                   "url": "https://lichess.org/@/bo",
```
The correct url is https://lichess.org/@/b.

AttackingOrDefending commented:
When the game ends I get:
```
[07/26/22 11:32:43][07/26/22 11:32:43]  INFO    INFO     ---  --- Process Queue. Total Queued: https://lichess.org/qyHkhDRJ/white0 Game over                        . Total Used: lichess-bot.py0:                       446lichess-bot.py
:155
```
or
```
                   [07/26/22 11:31:16]  INFO    INFO     ---  --- Process Queue. Total Queued: https://lichess.org/UG9Qp2DP/white0 Game over                        . Total Used: lichess-bot.py0:                       446lichess-bot.py
:155
```
![2](https://user-images.githubusercontent.com/40605232/180963466-dba4b5a0-d62b-46ba-89c0-8110a3068024.jpg)
![3](https://user-images.githubusercontent.com/40605232/180963479-07e170e9-d933-4858-982f-bae923b87098.jpg)

MarkZH commented:
Those logging messages are very strange. It's two messages mashed together, as if separate processes tried to write to the terminal simultaneously and they got interleaved. Does this happen after every game? Are you on Linux or Mac? I don't see this on my Windows machine.

AttackingOrDefending commented:
I am using Windows but I use the cmd. It happens every game. After downloading Terminal, which is what you seem to be using, I didn't get this problem either on powershell or on cmd. After running powershell (not from terminal), I got this problem again. So, this seems to be a problem of cmd and powershell only if you aren't using terminal. The version of my cmd and my terminal cmd are the same.
Powershell:
![1](https://user-images.githubusercontent.com/40605232/181312524-663880bb-e95c-455a-8af1-4030ae09bc31.jpg)

MarkZH commented:
On a hunch, I've moved the logging command back to its original location in the `start()` function. The nested function now just creates the text string to be logged. Let me know if this solves the problem.

AttackingOrDefending commented:
The problem is still there but the message has changed.
cmd:
![2](https://user-images.githubusercontent.com/40605232/181508958-ab855f21-aa4b-4196-b296-c53fedd234bf.jpg)

MarkZH commented:
I can't figure out how to run cmd without Terminal in Windows 11. So, some questions:

1. Does moving the `log_proc_count()` function outside of `start()` change anything (say from line 154 to line 118)?
2. If 1 doesn't work, does copying the returned string into that `logger` call eliminate the problem? I.e.,
```python
logger.info(log_proc_count("Freed", queued_processes, busy_processes))
```
becomes
```python
logger.info(f"+++ Process Freed. Total Queued: {queued}. Total Used: {used}")
```

AttackingOrDefending commented:
Both didn't work. After replacing all the lines with `log_proc_count` to `logger.info(f"+++ Process Freed. Total Queued: {queued_processes}. Total Used: {busy_processes}")` the problem is still there, so it doesn't seem to be a problem with `log_proc_count`.

MarkZH commented:
OK. I've uninstalled Terminal and now I'm seeing what your seeing, and not just when a game ends. This screenshot shows a challenge coming in the middle of another game.

![Screenshot 2022-07-28 233801](https://user-images.githubusercontent.com/10629050/181581236-f5ebf24b-267a-45ac-83de-5a2084622eba.png)

So, the console printer is having problems with multiple threads writing at the same time. There's nothing like this in the file logger.

Actually, this happens on the master branch as well.
![master branch](https://user-images.githubusercontent.com/10629050/181583604-5f30da22-4702-4623-a1bd-0a97b825aa0a.png)

Can you try the master branch and see if this happens for you, as well?

AttackingOrDefending commented:
After replacing
```python
        pool.apply_async(play_game,
                                     play_game_args,
                                     error_callback=game_error_handler)
```
with
```python
             pool.apply_async(play_game,
                                         [li,
                                          game_id,  # will hold the game id
                                          control_queue,
                                          user_profile,
                                          config,
                                          challenge_queue,
                                          correspondence_queue,
                                          logging_queue,
                                          game_logging_configurer,
                                          logging_level],
                                         error_callback=game_error_handler)
```
it worked correctly.

AttackingOrDefending commented:
> Can you try the master branch and see if this happens for you, as well?

It also happens at the master branch.

Since this seems to be unrelated with this PR, do you want to reverse the [last commit](https://github.com/ShailChoksi/lichess-bot/pull/506/commits/9ffde398fd6f87a7b91a59c8328e5143805b8748) and merge this, or do you want to add the fix to this PR?

MarkZH commented:
Inlining the argument list didn't fix it for me either. If it had, I would have been very annoyed.

I've reverted the `loc_proc_count()` commit, merged the latest changes to master, and added a few more small changes to logging exceptions. Since it's after midnight here, wait on merging. Tomorrow or the next day, I'll look all this over one more time and let you know that it's good to merge.

MarkZH commented:
This is now ready to merge.
